### PR TITLE
Add support for CSTC40 keyboard

### PR DESCRIPTION
**Pull request by felipecrp**
I recently bought this keyboard and found that @itsvar8 has reverse-engineered this keyboard's board and created a pull request to the QMK-VIAL project (https://github.com/vial-kb/vial-qmk/pull/517).

I backported the pull request to QMK and tested the firmware. I also fixed an RGB driver issue and added VIA support.

I wish to incorporate this keyboard into QMK to enable other users to customize their own keyboard.

Thank you, @itsvar8, for reverse-engineered the board, and thank you, great people of QMK Discord Channel, for helping me with some of the issues I handled in this PR.

## Description

This PR includes the CSTC40 keyboard for QMK.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/kprepublic/cstc40/single_pcb/rules.mk [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Build Options
+LTO_ENABLE = no
+TRI_LAYER_ENABLE = yes
```
Comment (waffle87) [inline keyboards/kprepublic/cstc40/single_pcb/rules.mk:3]:
```suggestion
# This file intentionally left blank
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/kprepublic/cstc40/daughterboard/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * |      |      |      |      |      |             |      |      |      |      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+[3] = LAYOUT_planck_mit(
+    _______,  QK_BOOT,  DB_TOGG,  RGB_TOG,  RGB_MOD,  RGB_HUI,  RGB_HUD,  RGB_SAI,  RGB_SAD,  RGB_VAI,  RGB_VAD,  KC_DEL,
+    _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,
+    _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,
+    _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______   
+)
+
```
[the remark this answers is not in the document]
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/daughterboard/keymaps/default/keymap.c:93]: The TRI_LAYER_ENABLE option manages to access layers 1, 2, and 3 by using the raise and lower layer keys. https://github.com/qmk/qmk_firmware/blob/master/docs/feature_tri_layer.md

File: keyboards/kprepublic/cstc40/daughterboard/keymaps/via/keymap.c [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * |      |      |      |      |      |             |      |      |      |      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+[3] = LAYOUT_planck_mit(
+    _______,  QK_BOOT,  DB_TOGG,  RGB_TOG,  RGB_MOD,  RGB_HUI,  RGB_HUD,  RGB_SAI,  RGB_SAD,  RGB_VAI,  RGB_VAD,  KC_DEL,
+    _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,
+    _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,
+    _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______   
+)
+
```
[the remark this answers is not in the document]
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/daughterboard/keymaps/via/keymap.c:93]: The TRI_LAYER_ENABLE option manages to access layers 1, 2, and 3 by using the raise and lower layer keys. https://github.com/qmk/qmk_firmware/blob/master/docs/feature_tri_layer.md

File: keyboards/kprepublic/cstc40/daughterboard/rules.mk [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Build Options
+LTO_ENABLE = no
+TRI_LAYER_ENABLE = yes
```
[the remark this answers is not in the document]
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/daughterboard/rules.mk:3]: The TRI_LAYER_ENABLE is needed for the current default layout. I will remove the LTO_ENABLE.

[the remark this answers is not in the document]
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/daughterboard/rules.mk:3]: I will move TRI_LAYER_ENABLE option to the keymap level.

**Comment by felipecrp** [comment]
> Since these two PCBs both use the `planck_mit` layout, you could remove a fair amount of duplication by structuring the keyboard folder as follows:
> 
> ```
> kprepublic/
> └── cstc40/
>     ├── single_pcb/
>     │   ├── single_pcb.c
>     │   ├── info.json
>     │   └── rules.mk
>     ├── daughterboard/
>     │   ├── daughterboard.c
>     │   ├── info.json
>     │   └── rules.mk
>     ├── keymaps/
>     │       ├── default/
>     │       │   └── keymap.c
>     │       └── via/
>     │           ├── keymap.c
>     │           └── rules.mk
>     ├── config.h
>     ├── info.json
>     └── README.md
> ```

Thank you for the suggestion. I just adjusted the code.

File: keyboards/kprepublic/cstc40/config.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define I2C1_SDA_PIN B9
+#define I2C1_SCL_PIN B8
+#define DRIVER_ADDR_1 0b1010000
+#define DRIVER_COUNT 1
+#define DRIVER_1_LED_TOTAL 47
+#define RGB_MATRIX_LED_COUNT DRIVER_1_LED_TOTAL
```
[the remark this answers is not in the document]
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/config.h]: Fixed. Thank you for the suggestion.

File: keyboards/kprepublic/cstc40/single_pcb/rules.mk (hunk 1 above)
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/single_pcb/rules.mk]: Fixed with the same comment of the previous suggestion.

File: keyboards/kprepublic/cstc40/info.json [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                { "matrix": [3, 5], "w": 2, "x": 5, "y": 3 },
+                { "matrix": [3, 7], "x": 7, "y": 3 },
+                { "matrix": [3, 8], "x": 8, "y": 3 },
+                { "matrix": [3, 9], "x": 9, "y": 3 },
+                { "matrix": [3, 10], "x": 10, "y": 3 },
+                { "matrix": [3, 11], "x": 11, "y": 3 }
+            ]
+        }
+    },
+	    "rgb_matrix": {
```
Comment (drashna) [inline keyboards/kprepublic/cstc40/info.json:77]:
```suggestion
    "rgb_matrix": {
```

File: keyboards/kprepublic/cstc40/keymaps/via/rules.mk [hunk 7]
```diff
@@ -0,0 +1,2 @@
+TRI_LAYER_ENABLE = yes
```
Comment (drashna) [inline keyboards/kprepublic/cstc40/keymaps/via/rules.mk:1]: Via enables this by default, so it's not necessary, here. 
```suggestion
```

Comment (felipecrp) [reply keyboards/kprepublic/cstc40/keymaps/via/rules.mk]: Thank you for the information.

File: keyboards/kprepublic/cstc40/info.json (hunk 6 above)
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/info.json]: Nice catch!

**Comment by felipecrp** [comment]
> Is there some reason you aren't enabling all RGB effects? That means Configurator users won't be able to switch to them.

I don't know. I just backported this firmware from a pull request on the qmk-vial (https://github.com/vial-kb/vial-qmk/pull/517) and made the modifications suggested by the reviewers. If you can advisor me, I can try to enable other effects.

**Comment by felipecrp** [comment]
> > I don't know. I just backported this firmware from a pull request on the qmk-vial ([vial-kb/vial-qmk#517](https://github.com/vial-kb/vial-qmk/pull/517)) and made the modifications suggested by the reviewers. If you can advisor me, I can try to enable other effects.
> 
> To enable additional effects, uncomment more of the `#define ENABLE_RGB_MATRIX_*` lines in the `config.h` file.

Thank you for the suggestion. Just included all RGB effects available at https://github.com/qmk/qmk_firmware/blob/master/docs/feature_rgb_matrix.md

File: keyboards/kprepublic/cstc40/daughterboard/daughterboard.c [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+#ifdef RGB_MATRIX_ENABLE
+const is31_led g_is31_leds[RGB_MATRIX_LED_COUNT] = {
+    { 0, L_1,  K_1,  J_1 },
+    { 0, L_2,  K_2,  J_2 },
+    { 0, L_3,  K_3,  J_3 },
+    { 0, L_4,  K_4,  J_4 },
+    { 0, L_5,  K_5,  J_5 },
+    { 0, L_6,  K_6,  J_6 },
+	{ 0, L_7,  K_7,  J_7 },
```
Comment (waffle87) [inline keyboards/kprepublic/cstc40/daughterboard/daughterboard.c:26]:
```suggestion
    { 0, L_7,  K_7,  J_7 },
```

File: keyboards/kprepublic/cstc40/info.json [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    "diode_direction": "ROW2COL",
+	"debounce": 2,
+    "features": {
+        "bootmagic": true,
+        "command": false,
+        "console": false,
+        "extrakey": true,
+        "mousekey": true,
+        "nkro": true,
+		"rgb_matrix": true
```
Comment (waffle87) [inline keyboards/kprepublic/cstc40/info.json:15]:
```suggestion
        "rgb_matrix": true
```

File: keyboards/kprepublic/cstc40/info.json [hunk 10]
```diff
@@ -0,0 +1,130 @@
+{
+    "manufacturer": "kprepublic",
+    "keyboard_name": "cstc40",
+    "maintainer": "itsvar8",
+    "bootloader": "stm32-dfu",
+    "diode_direction": "ROW2COL",
+	"debounce": 2,
```
Comment (waffle87) [inline keyboards/kprepublic/cstc40/info.json:7]:
```suggestion
        "debounce": 2,
```

File: keyboards/kprepublic/cstc40/single_pcb/single_pcb.c [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+#ifdef RGB_MATRIX_ENABLE
+const is31_led g_is31_leds[RGB_MATRIX_LED_COUNT] = {
+    { 0, L_1,  K_1,  J_1 },
+    { 0, L_2,  K_2,  J_2 },
+    { 0, L_3,  K_3,  J_3 },
+    { 0, L_4,  K_4,  J_4 },
+    { 0, L_5,  K_5,  J_5 },
+    { 0, L_6,  K_6,  J_6 },
+	{ 0, L_7,  K_7,  J_7 },
```
Comment (waffle87) [inline keyboards/kprepublic/cstc40/single_pcb/single_pcb.c:26]:
```suggestion
    { 0, L_7,  K_7,  J_7 },
```

File: keyboards/kprepublic/cstc40/daughterboard/daughterboard.c (hunk 8 above)
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/daughterboard/daughterboard.c]: Thank you for this suggestion. I reviewed all the files to check and correct identation.

File: keyboards/kprepublic/cstc40/info.json (hunk 9 above)
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/info.json]: 👍

File: keyboards/kprepublic/cstc40/info.json (hunk 10 above)
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/info.json]: 👍

File: keyboards/kprepublic/cstc40/single_pcb/single_pcb.c (hunk 11 above)
Comment (felipecrp) [reply keyboards/kprepublic/cstc40/single_pcb/single_pcb.c]: 👍

**Comment by felipecrp** [comment]
One thing that I observed it that the RGB does not turn off after the PC shutdown. I already included the RGBLIGHT_SLEEP directive, but apparently did not solved. Do you know what  should I do to turn off RGB on PC shutdown?

**Review by drashna — APPROVED** [review]

**Comment by felipecrp** [comment]
Hi, @waffle87! Could you please check whether the changes I made have correctly addressed your suggestions?

**Review by waffle87 — APPROVED** [review]

**Comment by itsvar8** [comment]
Good job, it's funny because I did every change requested too on my vial pr
and it's still sitting there

Il gio 12 ott 2023, 22:31 jack ***@***.***> ha scritto:

**Pull request merged**